### PR TITLE
Whack-a-mole scenario

### PR DESCRIPTION
**Pull request by kostmo**
image

### CODE REVIEW

File: test/integration/Main.hs [hunk 1]
```diff
@@ -162,6 +162,7 @@ testScenarioSolution _ci _em =
         , testSolution (Sec 5) "Challenges/2048"
         , testSolution (Sec 3) "Challenges/word-search"
         , testSolution (Sec 3) "Challenges/ice-cream"
+        , testSolution (Sec 90) "Challenges/gopher"
```
Comment (kostmo) [inline test/integration/Main.hs:165]: As is, this test take 45 seconds on my computer to complete.  So I'm not sure whether we actually want this to be part of the test suite.

Comment (kostmo) [reply test/integration/Main.hs]: I have now fixed this by adding a `wait 64` in the "polling" loop of the built robots.

File: data/scenarios/Challenges/_gopher/solution.sw [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            return $ x != "flower";
+        });
+
+    if shouldContinue {
+        if (n > 0) {
+            scanDirections $ n - 1;
+        } {
+            // Inserting this "wait" speeds up the
+            // integration test significantly (more than 3x)
+            wait 64;
```
Comment (kostmo) [inline data/scenarios/Challenges/_gopher/solution.sw:55]: The integration test runs much faster with this.

**Review by byorgey — COMMENTED** [review]
Nice!  My solution (written without looking at yours) is below.  Looks like they are functionally pretty similar.

I think we definitely need to fix the race condition before merging.  On the other hand I don't have a strong opinion about the `teleport`, I just wanted to hear your thought process and why you designed it that way.

```
def mod = \a. \b. (a - b*(a/b)) end

def forever = \c. c ; forever c end

def x4 = \c. c;c;c;c end

def drill_mound = \d.
  h <- scan d;
  if (h == inr "mound") { drill d } {}
end

def gopher_turret =
  forever (
    x4 ( turn left; drill_mound forward );
    drill_mound down
  )
end

def idoN_rec : int -> int -> (int -> cmd unit) -> cmd unit = \k. \n. \c.
  if (k == n) {} {c k; idoN_rec (k+1) n c}
end

def idoN : int -> (int -> cmd unit) -> cmd unit = idoN_rec 0 end

def doN : int -> cmd unit -> cmd unit = \n. \c. idoN n (\_. c) end

def place_turret = \row. \col.
  if (mod (2*row + col) 5 == 0) { build {gopher_turret}; return () } {}
end

def carpet_row = \r.
  idoN 28 (\c. place_turret r c; move);
  turn back; doN 28 move; turn left; move; turn left
end

def carpet = idoN 19 carpet_row end
```

File: data/scenarios/Challenges/_gopher/gopher.sw [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    let offsetXunsigned = (leadDist / 2) + offsetXrand in
+    let offsetYunsigned = leadDist - offsetXunsigned in
+
+    offsetX <- randSign offsetXunsigned;
+    offsetY <- randSign offsetYunsigned;
+
+    let startX = newDestinationX + offsetX in
+    let startY = newDestinationY + offsetY in
+
+    teleport self (startX, startY);
```
[the remark this answers is not in the document]
Comment (byorgey) [reply data/scenarios/Challenges/_gopher/gopher.sw:67]: Currently it also "burrows" through water, mountains, etc... which looks a bit strange.

File: data/scenarios/Challenges/_gopher/gopher.sw [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                    return (3, "Close one!");
+                } {
+                    if (newCount < startingAmount - 2) {
+                        return (4, "OK, no more Mr. Nice Gopher!");
+                    } {
+                        if (newCount < startingAmount - 1) {
+                            return (5, "Bet you can't do that again!");
+                        } {
+                            if (newCount < startingAmount) {
+                                return (6, "Beginner's luck!");
```
Comment (xsebek) [inline data/scenarios/Challenges/_gopher/gopher.sw:92]: This `if` nesting looks a bit strange, maybe you could use:
```
def elif = \t. \then. \else. {if t then else} end
def else = \t. t end
```

**Review by xsebek — COMMENTED** [review]
Wow, this is a huge challenge, I am not sure I am ready to face that gopher. :smile:

File: data/scenarios/Challenges/_gopher/gopher.sw (hunk 3 above)
[the remark this answers is not in the document]
Comment (kostmo) [reply data/scenarios/Challenges/_gopher/gopher.sw:67]: My original motivation for `teleport`-ing was to make it harder to predict where the gopher would go next.  However, I have now got rid of the `antenna`, which should prevent the player from "stalking" the gopher via `meet` as he tunnels (not to say that would be easy, either).

I think now I can forego the teleport and make the movement more natural.

File: data/scenarios/Challenges/_gopher/gopher.sw [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+def waitWhileHere = \e. \remainingTime.
+    here <- isHere e;
+    if here {
+        wait 1;
+        if (remainingTime > 0) {
+            waitWhileHere e $ remainingTime - 1;
+        } {
+            // Pick up the "dropping"
+            grab;
```
[the remark this answers is not in the document]
Comment (kostmo) [reply data/scenarios/Challenges/_gopher/gopher.sw:112]: Fixed.

File: data/scenarios/Challenges/_gopher/gopher.sw (hunk 4 above)
Comment (kostmo) [reply data/scenarios/Challenges/_gopher/gopher.sw]: fixed

**Comment by kostmo** [comment]
I think this is ready now.

**Review by byorgey — APPROVED** [review]
Looks great!  I had to make a minor change to my solution since the type of `drill` changed but after that change it still worked, and I like the new non-teleporting gopher.

I especially like this challenge because it feels like it might just be possible if you type really fast but it quickly gets tedious/impossible and you start to wonder if there is a better way...

**Pull request merged**